enhancements/microshift/kubernetes-for-industrial-edge.md

```diff
@@ -0,0 +1,435 @@
+dependencies as well. We are working with the control plane team to
+find ways to mitigate this risk, including moving some code out of the
+OpenShift API server into controllers that would not depend on
+kubernetes at all.
+
+Building a different form factor downstream of OpenShift means it is
+possible for behaviors introduced in the dependencies to cause issues
+within MicroShift that are not caught upstream. After the basic rebase
+automation is working, we will be trying to construct a CI job that
+can be run on pull requests to the repositories containing MicroShift
+dependencies to test upstream code changes using the MicroShift
+end-to-end test job.
+
+The support lifetime of MicroShift is the same as the version of
+OpenShift on which it is based, therefore it will be critical to
+automate the component update process as much as possible to ensure
+that MicroShift stays as closely up to date as it can. The team is
+developing a [rebase
+script](https://github.com/openshift/microshift/blob/main/scripts/rebase.sh)
+which will be used by an automated job to propose update pull
+requests. We anticipate periods of time when OpenShift is itself
+rebasing to a new version of kubernetes where MicroShift's
+dependencies are not compatible and cannot be updated. We will need to
+work with the rebase team to ensure those periods of time are as short
+as possible.
+
+Because MicroShift will run as only a single node, for application
+availability we will recommend running two single-node instances that
+deploy a common application in active/active or active/passive mode
+and then using existing tools to support failover between those states
+when either host is unable to provide availability.
+
+Because MicroShift does not embed the operating system, its life-cycle
+is independent of the OS. The components built into or delivered with
+MicroShift will be version-locked together. We will need to test
+MicroShift with the specific OS versions that we intend to support,
+which will necessarily increase the support matrix during the period
+when both RHEL 8 and 9 are covered.
+
+Some of the components being compiled together into a single binary
+are not intended to be used that way and are not tested that way
+upstream of MicroShift. There is community support for MicroShift and
+other similar projects, such as [k3s](https://k3s.io), which should
+help us if we encounter problems with the technical direction of those
+dependencies.
+
+### Drawbacks
+
+
+## Design Details
+
+### Open Questions [optional]
+
+1. How can we best minimize the window of time when OCP components are
+   not using the same version of kubernetes?
+2. Can we further reduce the dependencies of MicroShift by making
+   changes in any of the upstream components?
+3. Given the embedded image references, how will we build an
+   OKD-equivalent version of MicroShift?
+
+### Test Plan
+
+We will run the relevant kubernetes and OpenShift compliance tests in
```

Comment:
related: we may need to submit discrete k8s conformance results for this form-factor to cncf.

/cc @bparees

Reply:
That makes sense but i am curious where the line gets drawn as to what changes require a unique conformance result?

e.g. we have:

SNO
Hypershift
CRC(openshift local now)
Traditional Standalone OCP
Compact Standalone OCP
Microshift
Various service offering form factors (ARO, ROSA, etc)
Arguably KCP in some sense, someday.

Yet up until now as i understand it we've just done conformance testing against a traditional 6 node standalone cluster running on a single cloud provider.

So what defines a unique distribution that needs its own conformance test result?   (I realize of the above distributions, microshift is the one most likely to have unique behavior that might not be conformant, so for our own purposes of ensuring we didn't break stuff, running it makes sense)

Reply:
i suspect a unique installation flow is probably the best guiding input.

we submit for ocp (standalone), dedicated (rosa/aro/gcp models), and roks (ibm hosted).

Reply:
I added a note about this.